aead/src/lib.rs

```diff
@@ -0,0 +1,133 @@
+    /// a plaintext to accommodate this cipher's output.
+    type CiphertextOverhead: ArrayLength<u8>;
+
+    /// Construct a new stateful instance for the given key.
+    fn new(key: GenericArray<u8, Self::KeySize>) -> Self;
+
+    #[cfg(any(feature = "alloc", feature = "std"))]
+    /// Consume a plaintext vector, encrypt it in-place, and return the
+    /// resulting ciphertext.
+    fn encrypt_vec(
+        &mut self,
+        additional_data: &[u8],
+        nonce: &GenericArray<u8, Self::NonceSize>,
+        plaintext: Vec<u8>
+    ) -> Result<Vec<u8>, Error>;
+
+    #[cfg(any(feature = "alloc", feature = "std"))]
+    /// Consume a ciphertext vector, decrypt it in-place, and return a copy in
+    /// a new vector.
+    fn decrypt_vec(
+        &mut self,
+        additional_data: &[u8],
+        nonce: &GenericArray<u8, Self::NonceSize>,
+        ciphertext: Vec<u8>
+    ) -> Result<Vec<u8>, Error>;
+}
+
+/// A trait which can support a stateless RFC5116 authenticated encryption
+/// scheme. This is the standard RFC algorithm.
+pub trait StatelessAead {
+    /// The key size in a new method.
+    type KeySize: ArrayLength<u8>;
+    /// The length of a nonce.
+    type NonceSize: ArrayLength<u8>;
+    /// The maximum length of the nonce.
+    type TagSize: ArrayLength<u8>;
+    /// The amount of suffix padding, in bytes, which needs to be
+    /// appended to a plaintext to accomodate this cipher's output.
+    type CiphertextOverhead: ArrayLength<u8>;
+
+    #[cfg(any(feature = "alloc", feature = "std"))]
+    fn encrypt_vec(
+        key: &GenericArray<u8, Self::KeySize>,
+        additional_data: &[u8],
+        nonce: &GenericArray<u8, Self::NonceSize>,
+        plaintext: Vec<u8>
+    ) -> Result<Vec<u8>, Error>;
+
+    #[cfg(any(feature = "alloc", feature = "std"))]
+    fn decrypt_vec(
+        key: &GenericArray<u8, Self::KeySize>,
+        additional_data: &[u8],
+        nonce: &GenericArray<u8, Self::NonceSize>,
+        ciphertext: Vec<u8>
+    ) -> Result<Vec<u8>, Error>;
+}
+
+/// A wrapper structure to allow use of a stateless AEAD through the stateful
+/// interface.
+pub struct Stateful<Algo: StatelessAead> {
+    key: GenericArray<u8, Algo::KeySize>,
+    _aead: PhantomData<fn() -> Algo>
+}
+
+impl<Algo: StatelessAead> Aead for Stateful<Algo> {
+    type KeySize = Algo::KeySize;
+    type NonceSize = Algo::NonceSize;
+    type TagSize = Algo::TagSize;
+    type CiphertextOverhead = Algo::CiphertextOverhead;
+
+    fn new(key: GenericArray<u8, Self::KeySize>) -> Self {
+        Self {
+            key,
+            _aead: PhantomData::default(),
+        }
+    }
+
+    #[cfg(any(feature = "alloc", feature = "std"))]
+    /// Consume a plaintext vector, encrypt it in-place, and return the
+    /// resulting ciphertext.
+    fn encrypt_vec(
+        &mut self,
+        additional_data: &[u8],
+        nonce: &GenericArray<u8, Self::NonceSize>,
+        plaintext: Vec<u8>
```

Comment:
I think it will be better to keep using `plaintext: &[u8]`.

Reply:
That forces an allocation+copy in all cases, if the Vec has spare capacity, it may be avoided.

Reply:
...but only if you have a `Vec<u8>` to begin with. If you have a byte slice, it forces you to copy it into a `Vec<u8>` first.

I'd +1 @newpavlov that `&[u8]` would make more sense here.

Reply:
But your API will force `memmove` for schemes which place tag in the beginning of the ciphertext and will be less convenient if message is represented as slice. So I think we better to use slice for plaintext, also it will be consistent with the `BlockMode` trait.

Reply:
Any objection to doing `encrypt_vec` and `encrypt_to_vec`? The latter can chain to the former using the CiphertextOverhead length.